03-GettingStarted/06-http-streaming/README.md

```diff
@@ -0,0 +1,427 @@
+
+```python
+@mcp.tool(description="A tool that sends progress notifications")
+async def process_files(message: str, ctx: Context) -> TextContent:
+    await ctx.info("Processing file 1/3...")
+    await ctx.info("Processing file 2/3...")
+    await ctx.info("Processing file 3/3...")
+    return TextContent(type="text", text=f"Done: {message}")
+```
+
+### Client-side: Receiving Notifications
+The client must implement a message handler to process and display notifications as they arrive.
+
+```python
+async def message_handler(message):
+    if isinstance(message, types.ServerNotification):
+        print("NOTIFICATION:", message)
+    else:
+        print("SERVER MESSAGE:", message)
+```
+
+To enable notifications, ensure your server uses a streaming transport (like `streamable-http`) and your client implements a message handler to process notifications.
+
+---
+
+## 5. Streamable HTTP Transport
+
+### What is Streamable HTTP?
+Streamable HTTP is a transport mechanism in MCP that uses HTTP POST requests and supports streaming notifications (such as progress updates) from the server to the client. It is designed for modern web and cloud environments.
+
+### How It Works
+- The client sends a request to the server using HTTP.
+- The server can send notifications (e.g., progress, logs) back to the client while processing the request.
+- The final response is sent when processing is complete.
+
+### Benefits Over SSE and stdio
+- **Better scalability**: Works well with load balancers and cloud deployments.
+- **Stateless and stateful modes**: Supports both, with resumability.
+- **Rich notifications**: Send progress, logs, and other events during processing.
+- **Standard HTTP**: Easier to integrate with existing infrastructure.
+
+### Implementation Details
+- Uses JSON or SSE for response formats.
+- Requires clients to implement a message handler to process notifications.
+- Supports both synchronous and asynchronous tool execution.
+
+### Example Use Cases
+- Long-running document processing with progress updates
+- Real-time AI inference with streaming logs
+- Multi-client collaborative tools
+
+---
+
+## 6. Security Considerations
+
+### Overview
+Security is critical when exposing MCP servers over HTTP. Streamable HTTP introduces new attack surfaces and requires careful configuration.
+
+### Key Points
+- **Origin Header Validation**: Always validate the `Origin` header to prevent DNS rebinding attacks.
+- **Localhost Binding**: For local development, bind servers to `localhost` to avoid exposing them to the public internet.
+- **Authentication**: Implement authentication (e.g., API keys, OAuth) for production deployments.
+- **CORS**: Configure Cross-Origin Resource Sharing (CORS) policies to restrict access.
+- **HTTPS**: Use HTTPS in production to encrypt traffic.
+
+### Best Practices
```

Comment:
add lead in sentences, try to balance this out with more paragraphs, a bullet list should enhance the message of a header but not replace normal text